### ISSUE TITLE
[FEATURE] Loading .DEH files as PWADs in Multiplayer

### ISSUE THREAD

**Issue by Acts19quiz** [odamex/odamex#815]
**Is your feature request related to a problem? Please describe.**
Odamex seems to currently lack loading .DEH files for multiplayer situations. That may seem incredulous, given that there are both -file and -deh commands that can be passed to odasrv.exe, but that is the situation I've found in my testing.

The situation stems from testing Chex Quest, chex.wad, in an online environment. When playing local single-player, Odamex by default looks for fraggle's chex.deh in the same directory that chex.wad was found, then auto-loads it. That works well enough for single-player. However, it's an inappropriate solution for online since you start making really bad assumptions about what clients have what files downloaded, or what files the server is using. Odamex currently doesn't do this.

That's fine since, as mentioned, it shouldn't do that. However, where the problems begin is that DEH files Odamex searches for and relies on to accurately reproduce certain IWADs gameplay can't be loaded online. 

When loading chex.deh as a file:

`cmd /K odasrv -waddir "C:\****\CHEXQ" -iwad "chex.wad" -file "chex.deh" -config "C:\****\Chex_Quest.cfg" -port 10700`

The following error occurs: 

```
adding C:\****\chex.deh
 (1782 lumps)
Warning: chex.deh not loaded, experience may differ from the original!
```

However, chex.deh is found and read, and gets listed again in the "=== Odamex Server Initialized ===" section. Playtesting reveals it's acting correctly, *however*, chex.deh isn't listed among the PWADs used, either in Doomseeker or OdaLaunch. This is dangerous since, like mentioned, there are assumptions being made about what clients do or don't have. This is important for the case of Chex Quest since newcomers to Odamex browsing the server list looking for perspective games and servers to join certainly aren't going to notice, care, or perhaps even know about chex.deh if it's not listed in the required PWADs section and they're not prompted to download it.

Where things really go to Looney Town is when setting up the CFG for multi-WAD use. This is maybe where things cross over between a desired feature (clear DEH loading) and a bug in the way altering the map rotation alters DEH handling.

```
// --- Map Rotation (Based on the PSN version of Doom) --- 
addmap E1M1
addmap E1M2
addmap E1M3
addmap E1M4
addmap E1M5
```

Produces no particular result with or without CHEX2.WAD being loaded. **_However_**, using the multi-WAD format:

```
// --- Map Rotation (Based on the PSN version of Doom) --- 
addmap E1M1 chex.wad
addmap E1M2
addmap E1M3
addmap E1M4
addmap E1M5
```

makes things worse. Like before, chex.deh is listed in the initial loading, and there is the warning that the "experience may differ". However, chex.deh fails to show up in the "Odamex Server Initialized" section, and playtesting confirms it isn't being deployed.

Changing -file "chex.deh" to -deh "chex.deh" makes things worse. Even with CHEX.WAD factored out, chex.deh only shows loading in the first section. The "experience" warning isn't displayed there, but it is int he "Odamex Server Initialized" section. chex.deh is also absent there. Playtesting confirms its lack of load.

TL;DR: There's no easy way to distribute the orignal Chex Quest experience online, no easy way to load DEH files as PWADs, and no way to communicate to random game seekers what DEH files are actually being used and required by the server.

**Describe the solution you'd like to see**
This seems like a complicated problem with a mixture of bugs and presentation issues from a lack of support or features realted to multiplayer DEH loading. The simplest way to boil it down is:

1. Find out why DEH loading is impacted by multi-wad loading in server CFGs.
2. Treat a la carte DEH files as PWADs.
3. Make sure any DEHs loaded get displayed as PWADs in OdaLaunch, and properly communicated to the server browser of choice. I have no idea how ZDaemon handles this, but I know Zandronum does it this way.

**Describe alternatives you've considered**
This... is pretty 'out there' and might be considered outside of Odamex's scope. However, one alternative solution is to discard chex.deh altogether and instead handle it the way ZDoom did.

Regarding LANGUAGE: Upon Chex Quest detection (either chex.wad, or in ZDoom's case, chex3.wad), a number of Chex Quest-specific tokens are used. E.g., instead of GOTPLASMA, a GOTPHASINGZORCHER token is used. Odamex already has the ZDoom-style LANGUAGE lump, so this might not be too much of a stretch. This is sometimes mixed with special LANGUAGE formatting, like, "$ifgame(chex) QSPROMPT =".

The more complicated arm of this would be coding in specific behaviour for Chex Quest actors and other situations. As in, if the game is detected to be Chex, alter the Zombie actors to not drop ammo and only attack in melee range. Though I should note that this is a more complicated solution to fix a DEH loading error, one that wouldn't help other DEH games like HacX.

**Additional context**
[Here is a test package](https://github.com/odamex/odamex/files/10214633/DEH_test.zip) to drop into the base directory with odamex.exe. On Windows, just run Chex_launch_10666.bat. It sets up on port 10666 with "sv_usemasters 0". For convenience's sake, also grab chex.wad and CHEX2.WAD from a place like below and place them in with odamex.exe.

[http://ifocserv.net/ftp/wads/chex.wad](http://ifocserv.net/ftp/wads/chex.wad)
[http://ifocserv.net/ftp/wads/chex2.wad](http://ifocserv.net/ftp/wads/chex2.wad)

**Comment by rakohus** [comment]
Changed to bug status instead of feature.

**Comment by electricbrass** [comment]
I think part of this is not actually a bug. When adding maps to the maplist, chex.deh needs to be included in the wad list.

i.e.
the config you gave should be:
```
addmap E1M1 chex.wad chex.deh
addmap E1M2
addmap E1M3
addmap E1M4
addmap E1M5
addmap E1M1 chex.wad chex.deh CHEX2.WAD
addmap E1M2
addmap E1M3
addmap E1M4
addmap E1M5

set sv_usemasters 0
```

When this is done, it does appear as a listed file in Odalaunch and Doom Explorer.
deh files aren't loading for each map in the maplist without specifying for the same reason that PWADs also must be specified.

It appears the issues is that dehacked patches are only correctly loaded when using `-deh`, and not when using the `wad` command or anything else such as the maplist.

**Comment by bcahue** [comment]
Is this still an issue?

**Comment by electricbrass** [comment]
Pretty sure this is fixed
